File: 0046-Permutations/lc0046.hs
-- Haskell solution.

permute :: [a] -> [[a]]
permute []     = [[]]
permute (x:xs) = (permute xs) >>= (ins x)

ins :: a -> [a] -> [[a]]
ins x []     = [[x]]
ins x (y:ys) = [x:y:ys] ++ (map (y:) (ins x ys))

File: 0046-Permutations/lc0046.py
# Python solution.

class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        return reduce(lambda ll, x: [l[:i] + [x] + l[i:] for l in ll for i in range(len(l)+1)],
                      nums,
                      [[]])
